homeassistant/helpers/trigger_template_entity.py

```diff
@@ -231,16 +255,22 @@ def _process_manual_data(self, value: Any | None = None) -> None:
         Ex: self._process_manual_data(payload)
         """
 
-        self.async_write_ha_state()
+        run_variables: dict[str, Any] = {"value": value}
         this = None
         if state := self.hass.states.get(self.entity_id):
             this = state.as_dict()
-
-        run_variables: dict[str, Any] = {"value": value}
         # Silently try if variable is a json and store result in `value_json` if it is.
         with contextlib.suppress(*JSON_DECODE_EXCEPTIONS):
             run_variables["value_json"] = json_loads(run_variables["value"])
         variables = {"this": this, **(run_variables or {})}
+        self._render_availability_template(variables)
+
+        self.async_write_ha_state()
+        this = None
+        if state := self.hass.states.get(self.entity_id):
+            this = state.as_dict()
+
+        variables["this"] = this
```

Comment:
I think the existing code is strange, why do we unconditionally call `self.async_write_ha_state` before rendering any templates. I'm guessing it has to do with startup behavior, before the entity is first added. However, the solution to that is to use `TemplateStateFromEntityId`, as is done in `TemplateEntity`: https://github.com/home-assistant/core/blob/d9eada044ccf28b6d8ebea0db637cae3b827ccc7/homeassistant/components/template/template_entity.py#L350

I would suggest we clean that up in an initial PR, then rebase this PR after.

Reply:
I also think the similar existing code in the template integration's `TriggerEntity` should be changed to use `TemplateStateFromEntityId` here: https://github.com/home-assistant/core/blob/d9eada044ccf28b6d8ebea0db637cae3b827ccc7/homeassistant/components/template/trigger_entity.py#L44-L48

Reply:
> I think the existing code is strange, why do we unconditionally call self.async_write_ha_state before rendering any templates.

Because otherwise `this` would refer to the previous state instead of the current state.
With the current suggestion it's the same thing, but now it's moved to https://github.com/home-assistant/core/pull/128054/files#diff-414fea43d4fb87d5367be0a224750dbd804079ab4da80743243af0c66808e962R268 instead (so it considers the availability prior which is the issue to solve)

Reply:
> Because otherwise this would refer to the previous state instead of the current state.

That's how it's supposed to work. The `this` object is supposed to make it easy to access the current state when calculating a new state: https://github.com/home-assistant/core/pull/65201

Reply:
Related PR to handle edge case for new entity: https://github.com/home-assistant/core/pull/70911